Hdoan/add git pr chain new command

hdoan741 opened the pull request:
In this PR I introduce a few new features:
- A new command `git-pr-chain new` to automagically create a `git-pr-chain:` token in a commit. 
- A new command `git-pr-chain end-chain` to add an empty "git-pr-chain: STOP" commit
- Use github user name as the default prefix, if none is provided in gitconfig. This will make it easier for new people to adopt (1 less thing to config)

The `new` serves 2 purposes:
- Workflow wise, I create commits before thinking about making branches or marking them as the start of a new PR. This command allows me to easily go back and add `git-pr-chain` token to a commit.
- The chain name is irrelevant and can be generated from the commit message instead. One less thing I have to think about

It doesn't affect the existing workflow. People who want to add `git-pr-chain` to their commits manually can still do that.<git-pr-chain>

Too bad I cannot use git-pr-chain to open a pull request to the original repo directly...

#### Commits in this PR
1. add git-pr-chain new command
1. fix remote url parser
1. add a new command to mark the end of a chain
1. use user login name as the default prefix

Commit:
add git-pr-chain new command

git-pr-chain: add_git_pr_chain_new_command

Commit:
add function to generate new branch name & update commit msg

Commit:
fix remote url parser

git-pr-chain: fix_remote_url_parser

Commit:
add a new command to mark the end of a chain

git-pr-chain: add_a_new_command_to_mark_the_end_of_a_c

Commit:
update msg

Commit:
make the new command dumb

jlebar commented on git-pr-chain.py:
```diff
@@ -166,17 +166,28 @@ def __init__(self, sha: str, parent: Optional["Commit"]):
+
+    @cached_property
+    @traced
+    def parse_pr_chain(self) -> Optional[str]:
```
If it's a property, then the name shouldn't be a verb.

jlebar commented on git-pr-chain.py:
```diff
@@ -166,17 +166,28 @@ def __init__(self, sha: str, parent: Optional["Commit"]):
+    @cached_property
+    @traced
+    def parse_pr_chain(self) -> Optional[str]:
+        """Return pr chain if it exists in this PR, none otherwise."""
```
You're probably more familiar than I am with the code at this point, but is "a commit's PR chain" an existing concept?  Perhaps we should say, a commit's "git-pr-chain annotation"?

Similarly, this object is a commit, so presumably it's not "if it exists *in this PR*"?

hdoan741 replied:
Yeah, I'm struggling to find the right terminology, so please feel free to suggest good nouns to describe thing. I think pr_chain_annotation sounds good to me.

jlebar commented on git-pr-chain.py:
```diff
@@ -700,6 +715,37 @@ def cmd_merge(args):
+def cmd_new_pr(args):
+    # Use this command to mark the top commit as a new PR in the chain. It will
+    # generate the PR branch automatically
+    def generate_pr_chain_name(commit_title) -> str:
```
generate_branch_name?  or generate_pr_chain_annotation?

jlebar commented on git-pr-chain.py:
```diff
@@ -700,6 +715,37 @@ def cmd_merge(args):
 
+def cmd_new_pr(args):
+    # Use this command to mark the top commit as a new PR in the chain. It will
+    # generate the PR branch automatically
```
Add docstring

jlebar commented on git-pr-chain.py:
```diff
@@ -700,6 +715,37 @@ def cmd_merge(args):
+        # Strip _ away from the head and tail: ___xx_xx___ -> xx_xx
+        strip_start_end_underscore = re.sub(r"^_*|_*$", "", truncate_continuous_underscore)
+        # Only keep 40 characters
+        return strip_start_end_underscore[0:40]
```
I'm wondering if we should add a nonce to this (probably sufficient to generate 64 random bits) so that branch names don't collide.  The first line of a commit isn't necessarily unique...

jlebar replied:
This is the only thing that's keeping me from approving this as-is.

jlebar commented on git-pr-chain.py:
```diff
@@ -700,6 +715,37 @@ def cmd_merge(args):
+        # Only keep 40 characters
+        return strip_start_end_underscore[0:40]
+
+    # 1. Get current commit message and verify that it doesn't have git-pr-chain name on it
```
80 chars (run `black` over this file?)

jlebar replied:
s/name/annotation/

jlebar commented on git-pr-chain.py:
```diff
@@ -700,6 +715,37 @@ def cmd_merge(args):
+
+    maybe_pr_chain = head_commit.parse_pr_chain
+    if maybe_pr_chain:
+        fatal(f"There is already a git-pr-chain for this commit: {maybe_pr_chain}")
```
s/for/annotation on/

jlebar commented on git-pr-chain.py:
```diff
@@ -700,6 +715,37 @@ def cmd_merge(args):
+        fatal(f"There is already a git-pr-chain for this commit: {maybe_pr_chain}")
+
+    # 2. Create a name and update the commit message
+    print(f"Use {generate_pr_chain_name(head_commit.commit_title)} as the brach name")
```
s/Use/Using/

jlebar commented on git-pr-chain.py:
```diff
@@ -700,6 +715,37 @@ def cmd_merge(args):
+    if maybe_pr_chain:
+        fatal(f"There is already a git-pr-chain for this commit: {maybe_pr_chain}")
+
+    # 2. Create a name and update the commit message
```
Update "name" per above

jlebar commented on git-pr-chain.py:
```diff
@@ -710,6 +756,9 @@ def main():
     sp_log = subparser.add_parser("log", help="List commits in chain")
     sp_log.set_defaults(func=cmd_log)
 
+    sp_new = subparser.add_parser("new", help="Add a new PR to the chain")
```
"Mark HEAD as starting a new PR in the chain"?

jlebar reviewed (commented):
Thank you for the PR!

Update the README to describe the new command?

Commit:
update the term used for git-pr-chain annotation

hdoan741 commented:
@jlebar I added all my changes to this PR and addressed most of your comments. I'll update the README later

jlebar commented on git-pr-chain.py:
```diff
@@ -750,6 +750,21 @@ def generate_pr_chain_annotation(commit_title) -> str:
+    chain_end_msg = "=== git-pr-chain: STOP ==="
+    print(f"Create empty commit to stop the chain: ", chain_end_msg)
+    if not DRY_RUN:
+        git("commit", "--allow-empty", "-m", chain_end_msg)
```
For this command, since it's just `git commit --allow-empty -m "git pr-chain: STOP` -- do we really need a command in the tool for it?  Like, if you know how to use interactive rebase, which is kind of a prerequisite for using this at all, then should we assume you also know how to type that, or add an alias of your own?

I guess I lean towards having fewer commands so you don't have to learn / understand as much in order to use the tool.

hdoan741 replied:
I think commands are abstractions... I want them to capture the intention of users, ie what they want to do, and simplify it for them. And this command capture that intention, and being convenient enough that you don't need to think about how it works behind the scene.

hdoan741 replied:
I think `end-chain` is an important enough concept that the tool should support it formally. Also, by making it an abstraction (a command), we can hide or change the implementation however we want.

jlebar replied:
I think you do need to think about how it works behind the scenes, though, because you're going to have to delete this commit to "un-stop" your chain.  You have to understand that deleting the commit is tantamount to un-stopping things, that there's no other metadata you need to delete.

hdoan741 replied:
Yeah, I agree user still needs to understand the mechanism of "stopping the chain" (and we should document that to unstop the chain, remove the commit added.) On `git rebase -i` flow, it will look like this, and I think it's pretty clear what'll happen:
![Screen Shot 2020-10-21 at 11 08 09 AM](https://user-images.githubusercontent.com/687617/96760835-dc2a2c80-138d-11eb-81e8-2e263d7dcd3e.png)

Anyway, I'm an advocator for command-line aliases and ease of use, so I still find value in having this command. Even when it's something users can do manually, it's quite a lot to remember or to type to achieve something like this. Another benefit is that it'll show up on the command's help page. So when someone looks for a way to "stop adding commits to a PR", they can see it from the command line help and don't need to look for help from the repo's documentation.

jlebar replied:
Okay, I guess that isn't so bad.

Apparently the behavior of git rebase has changed wrt empty commits?  Previously I remember it would drop them.  Now it seems that it will keep them around and annotate with `#empty`?  That's better for this use case!

jlebar commented on git-pr-chain.py:
```diff
@@ -268,7 +268,7 @@ def gh_branch_prefix():
     except subprocess.CalledProcessError:
         # git config exits with an error code if the config key is not found.
-        return ""
+        return gh_client().get_user().login + "/"
```
I'm fine with this change in principle, but as-is this breaks existing users, right?  If I have some existing PRs and I don't have a branch prefix, then this will change the branch names of all my outstanding PRs?

hdoan741 replied:
Yeah, you're right. It'll create new PRs for existing PR... In git-stacked-pr, we encode the full name in the commit message, including the PR link (once it's created), but git-pr-chain it's a separate config...

What do you think we can do to not breaking existing users' PR? I can separate this to its own PR if you want (so that the rest can go in)

jlebar replied:
It's a little complex, but we could add a "time bomb"?  Basically the first time it runs it remembers the date (you could save it in the gitconfig) and prints a warning that at date + 1 month it will change the default?

hdoan741 replied:
Sound like a good idea. I've removed this commit from the PR for now, an I can redo it, together with this change, in another PR.

Commit:
update readme

hdoan741 commented:
@jlebar I've added readme and removed the change to use GitHub's username by default. It's easy enough to set up the config. I can still send it out as a separate PR if you think it's worth having.

jlebar commented on README.md:
```diff
@@ -3,6 +3,13 @@
 `git-pr-chain` is an opinionated tool that manages dependent GitHub pull
 requests.
 
+Use `git-pr-chain` you will be able to:
```
s/Use/Using/

jlebar commented on README.md:
```diff
@@ -3,6 +3,13 @@
 
+Use `git-pr-chain` you will be able to:
+
+- ✅ Create multiple Github PRs with 1 local git branch (each with one or more commits)
```
s/1/one/ and below

jlebar commented on README.md:
```diff
@@ -3,6 +3,13 @@
+Use `git-pr-chain` you will be able to:
+
+- ✅ Create multiple Github PRs with 1 local git branch (each with one or more commits)
+- ✅ Show the list of depedent PRs on the PR summary automatically
```
s/Show ... on/Include ... in/

jlebar commented on README.md:
```diff
@@ -31,8 +38,10 @@ $ git commit -m "Add foo\ngit-pr-chain: add-foo"
 $ echo "blah blah" > foo
 $ git commit -a -m "Update foo"
 
+# We can also run `git-pr-chain new` to add PR-chain annotation automatically
```
s/to add PR-chain/to add the git-pr-chain/

jlebar replied:
s/automatically// (it's not automatic, you have to ask for it! :)

jlebar commented on README.md:
````diff
@@ -55,18 +64,58 @@ $ git-pr-chain merge --merge-method=rebase
+```
+    log                 List commits in chain
+    new                 Mark HEAD as starting a new PR in the chain
+    end-chain           Add a commit to mark the end of the chain
````
s/chain/chain.  No commits beyond this point will be uploaded to github./

jlebar commented on git-pr-chain.py:
```diff
@@ -710,6 +774,12 @@ def main():
+    sp_new = subparser.add_parser("new", help="Mark HEAD as starting a new PR in the chain")
+    sp_new.set_defaults(func=cmd_new_pr)
+
+    sp_end = subparser.add_parser("end-chain", help="Add a commit to mark the end of the chain")
```
Same comment as in the readme on this line

jlebar commented on git-pr-chain.py:
```diff
@@ -710,6 +774,12 @@ def main():
+    sp_new = subparser.add_parser("new", help="Mark HEAD as starting a new PR in the chain")
+    sp_new.set_defaults(func=cmd_new_pr)
+
+    sp_end = subparser.add_parser("end-chain", help="Add a commit to mark the end of the chain")
```
Please run `black` before submitting!

hdoan741 replied:
What does running "black" mean?

hdoan741 replied:
Found it. A code formatting tool for python! 😮

jlebar reviewed (commented):
A bunch of comments on the README, other than that the only outstanding item in the code for me is the nonce in the branch name.

Commit:
add hex padding to branch name

git-pr-chain: add_hex_padding_to_branch_name_efa2

Commit:
update documentation

Commit:
running black

hdoan741 commented on git-pr-chain.py:
```diff
@@ -701,6 +721,62 @@ def cmd_merge(args):
+            r"^_*|_*$", "", truncate_continuous_underscore
+        )
+        # Only keep 40 characters, padding 4 random hex characters at the end
+        random_padding = "".join(random.choice(string.hexdigits) for n in range(4))
+        return f"{strip_start_end_underscore[0:40]}_{random_padding.lower()}"
```
It'll now generate a branch name like. The 4 character padding should provide it enough uniqueness, I think.
```
    git-pr-chain: add_hex_padding_to_branch_name_efa2
```

jlebar replied:
I agree.

Just to show my work: 4 hex digits is 2^32 possible nonces.  To get a collision with probability P, you'd need roughly `n = sqrt(2^33 * P)` commits, https://en.wikipedia.org/wiki/Birthday_problem#Square_approximation.  Setting P = 0.1%, that works out to 2,930 PRs, all starting with the same 40 characters.

That's a lot.

hdoan741 reviewed (commented):
@jlebar I've addressed your feedback. Please let me know if there is any other changes you'd like me to make

jlebar reviewed (approved):
🎉

jlebar commented:
Ship it.  :)

hdoan741 commented:
@jlebar 💯  thank you. 
I think I don't have permission to merge the PR & you have to do it.

jlebar commented:
Heh, of course.

Merged!
